baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreferInputStreamTransferTo.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+import java.util.List;
+
+/**
+ *  Allow for optimization when underlying input stream (such as `ByteArrayInputStream`, `ChannelInputStream`) overrides
+ *  `transferTo(OutputStream)` to avoid extra array allocations and copy larger chunks at a time (e.g. allowing 16KiB
+ *  chunks via `ApacheHttpClientBlockingChannel.ModulatingOutputStream` from #1790).
+ * <p>
+ *  When running on JDK 21+, this also enables 16KiB byte chunk copies via `InputStream.transferTo(OutputStream)`
+ *  per JDK-8299336, where as on JDK < 21 and when using Guava `ByteStreams.copy` 8KiB byte chunk copies are used.
+ * <p>
+ * References:
+ *  <ul>
+ *  <li><a href="https://github.com/palantir/hadoop-crypto/pull/586">https://github.com/palantir/hadoop-crypto/pull/586</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8299336">https://bugs.openjdk.org/browse/JDK-8299336</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8067661">https://bugs.openjdk.org/browse/JDK-8067661</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8265891">https://bugs.openjdk.org/browse/JDK-8265891</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8273038">https://bugs.openjdk.org/browse/JDK-8273038</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8279283">https://bugs.openjdk.org/browse/JDK-8279283</a></li>
+ *  <li><a href="https://bugs.openjdk.org/browse/JDK-8296431">https://bugs.openjdk.org/browse/JDK-8296431</a></li>
+ *  </ul>
```

Comment:
Much of this would work nicely in the summary field — error-prone is fantastic at providing relevant information to devs at the moment it’s most useful :-)

Reply:
I can move a lot of this and some of `summary` to [`explanation`](https://errorprone.info/api/latest/com/google/errorprone/BugPattern.html#explanation())

Reply:
updated